File: doc/lua-examples/level.lua
--- Examples on how to use level objects from Lua

-- GRIB1 level

-- Create with leveltype followed by l1 and l2
-- l1 and l2 are optional, according to the level type

-- For example, surface level has no parameters
local o = arki_level.grib1(1)
ensure_equals(o.style, "GRIB1")
ensure_equals(o.type, 1)
ensure_equals(tostring(o), "GRIB1(001)")

-- For example, altitude above MSL has 1
local o = arki_level.grib1(103, 1000)
ensure_equals(o.style, "GRIB1")
ensure_equals(o.type, 103)
ensure_equals(o.l1, 1000)
ensure_equals(tostring(o), "GRIB1(103, 01000)")

-- For example, layer of altitude above MSL has 2
local o = arki_level.grib1(104, 123, 223)
ensure_equals(o.style, "GRIB1")
ensure_equals(o.type, 104)
ensure_equals(o.l1, 123)
ensure_equals(o.l2, 223)
ensure_equals(tostring(o), "GRIB1(104, 123, 223)")


-- GRIB2 single level

-- Create with type, scale and value
local o = arki_level.grib2s(103, 0, 1000)

-- Accessors
ensure_equals(o.style, "GRIB2S")
ensure_equals(o.type, 103)
ensure_equals(o.scale, 0)
ensure_equals(o.value, 1000)
ensure_equals(tostring(o), "GRIB2S(103, 000, 0000001000)")


-- GRIB2 layer

-- Create with type1, scale1, value1, type2, scale2, value2
local o = arki_level.grib2d(103, 0, 1000, 103, 0, 1100)

-- Accessors
ensure_equals(o.style, "GRIB2D")
ensure_equals(o.type1, 103)
ensure_equals(o.scale1, 0)
ensure_equals(o.value1, 1000)
ensure_equals(o.type2, 103)
ensure_equals(o.scale2, 0)
ensure_equals(o.value2, 1100)
ensure_equals(tostring(o), "GRIB2D(103, 000, 0000001000, 103, 000, 0000001100)")

File: python/arkimet/formatter/level.py
from arkimet.formatter import Formatter
from arkimet.formatter.eccodes import GribTable
import os


def format_level(v):
    if v["style"] == "GRIB1":
        type, l1, l2 = v["level_type"], v.get("l1"), v.get("l2")
        levels = GribTable.load(1, "3")
        if levels.has(type):
            if l1 is None:
                l1 = "-"
            if l2 is None:
                l2 = "-"
            return "{} {} {} {}".format(
                                 levels.abbr(type),
                                 levels.desc(type),
                                 l1, l2)
        else:
            return None
    elif v["style"] in ("GRIB2S", "GRIB2D"):
        # TODO: index centre, table_version, local_table_version
        centre, table_version, local_table_version = None, None, None

        if v["style"] == "GRIB2S":
            type1, scale1, value1 = v["level_type"], v["scale"], v["value"]
            type2, scale2, value2 = None, None, None
        else:
            type1, scale1, value1 = v["l1"], v["scale1"], v["value1"]
            type2, scale2, value2 = v["l2"], v["scale2"], v["value2"]

        levels = GribTable.load(
                2, os.path.join(GribTable.get_grib2_table_prefix(centre, table_version, local_table_version), "4.5"))

        def format_single_level(type, scale, value):
            if levels.has(type):
                if value is None:
                    value = "-"
                if scale is None:
                    scale = "-"
                return "{} {} {} {}".format(levels.abbr(type), levels.desc(type), scale, value)
            else:
                return None

        if type2 is None:
            return format_single_level(type1, scale1, value1)
        else:
            return format_single_level(type1, scale1, value1) + " " + format_single_level(type2, scale2, value2)


Formatter.register("level", format_level)
